prov/rxd: bug fixes, enable provider, add to Travis

aingerson opened the pull request:
Summary:
- enable provider by default
- AV return bug fixes
- peer connection organization/retrying of connection packets
- add retry warning message
- recv_cancel fix to maintain sequence numbers
- length calculation fix/clean
- add exclude file
- add UDP;ofi_rxd to make test for Travis

Also throwing a gitignore fix in here (add fabtests/* extension since repos were merged)

Updated:
Added commit to enable psm2;ofi_rxd by copying resolved address needed for the name server.
Added commit to re-enable skipping of utility providers in ubertest.

shefty commented on prov/rxd/src/rxd_av.c:
```diff
@@ -153,7 +154,7 @@ int rxd_av_insert_dg_addr(struct rxd_av *av, const void *addr,
 			     flags, context);
 	if (ret != 1)
-		return -errno;
+		return -FI_EINVAL;
```
Why isn't the return value simply passed through?

shefty commented on prov/rxd/src/rxd_av.c:
```diff
@@ -84,6 +84,7 @@ static int rxd_av_set_addrlen(struct rxd_av *av, const void *addr)
 	if (ret != 1) {
 		FI_WARN(&rxd_prov, FI_LOG_AV, "addr insert failed: %d (%s)\n",
 			-ret, fi_strerror(-ret));
+		ret = -FI_EINVAL;
```
Why not pass ret through?

aingerson replied:
The functions rxd_av_set_addrlen and rxd_av_insert_dg_addr (comment above) both deal with fi_av_insert (which would return 1 on success) and other functions that return error codes. So if we pass ret through on failure (ret = 0) then we can't catch the error.

shefty commented on prov/rxd/src/rxd.h:
```diff
@@ -192,6 +192,7 @@ struct rxd_ep {
 	struct dlist_entry rx_list;
 	struct dlist_entry rx_tag_list;
 	struct dlist_entry active_peers;
+	struct dlist_entry waiting_peers;
```
If you call this rts_sent_list, it would be clearer what the state of the entries are that are in the list and what they are waiting on.

shefty commented on prov/rxd/src/rxd_ep.c:
```diff
@@ -915,25 +936,11 @@ static void rxd_ep_progress(struct util_ep *util_ep)
 		goto out;
 
-	current = fi_gettime_ms();
+	dlist_foreach_container(&ep->waiting_peers, struct rxd_peer, peer, entry)
```
Note that this iterator does not handle items being removed remove the list.

shefty commented on prov/rxd/src/rxd_ep.c:
```diff
@@ -664,10 +670,7 @@ void rxd_ep_send_ack(struct rxd_ep *rxd_ep, fi_addr_t peer)
-	if (ret)
-		FI_WARN(&rxd_prov, FI_LOG_EP_CTRL, "Unable to send ack\n");
-
+	rxd_ep_retry_pkt(rxd_ep, pkt_entry);
```
If the return code isn't used anywhere else, the function can be converted to void

shefty commented on prov/rxd/src/rxd_cq.c:
```diff
@@ -214,7 +214,7 @@ static void rxd_complete_rx(struct rxd_ep *ep, struct rxd_x_entry *rx_entry)
 	     (rx_entry->cq_entry.flags & FI_RMA ||
 	     rx_entry->cq_entry.flags & FI_ATOMIC)) ||
-	     rx_entry->flags & RXD_NO_RX_COMP)
+	     rx_entry->flags & RXD_NO_RX_COMP || rx_entry->flags & RXD_CANCELLED)
```
This can be simplified as
```
rx_entry->flags & (RXD_NO_RX_COMP | RXD_CANCELED)
```
A similar change can be made to cq_entry.flags check above it.

shefty replied:
Btw, how does cancel differ from the no_rx_comp flag?

aingerson replied:
The cancel flag is checked in progress_op to increment the sequence numbers for the entire canceled message so we can account for them without having to process them while the no_rx_comp_flag still processes the message just without generating a completion.

aingerson commented:
Updated

aingerson commented:
@a-ilango see ubertest changes

aingerson commented:
@shefty psm2;ofi_rxd works fine because of the src_addr copy (included) but I'm omitting it from ubertest testing by re-enabling the utility provider skip. This includes an extra protocol exchange just to be sure and skip exchange (to avoid one side skipping psm2;ofi_rxd but the other one trying to use psm2 and waiting for the other side). Let me know if you have any more comments about these changes. They should pass the CI now.

a-ilango commented on fabtests/ubertest/uber.c:
```diff
@@ -328,6 +328,52 @@ static int ft_recv_test_info(void)
+	if (ret)
+		return ret;
+
+	return skip || remote_skip ? FT_SKIP : 0;
```
We should probably not return FT_SKIP. The caller assumes return value of 1 means skip.

a-ilango commented on fabtests/ubertest/uber.c:
```diff
@@ -328,6 +328,52 @@ static int ft_recv_test_info(void)
+		return ret;
+
+	if (info->ep_attr->protocol != remote_protocol)
+		skip = 1;
```
I guess there's no need to exchange skips once we find that there is a protocol mismatch.

a-ilango commented on fabtests/ubertest/uber.c:
```diff
@@ -328,6 +328,52 @@ static int ft_recv_test_info(void)
+	}
+
+	return 0;
+}
```
Add `ntohl` and `htonl` for the `uint32_t`s.

aingerson commented:
Ok... updated AGAIN. @a-ilango I decided that instead of skipping we should just return ENODATA since that's what the rest of the test does if the provider does not support the configuration. As a result, I removed all of the FI_EP_RDM test cases from the verbs config so it can pass the cases. Let me know your thoughts.

shefty commented on prov/rxd/src/rxd_av.c:
```diff
@@ -415,8 +418,11 @@ int rxd_av_create(struct fid_domain *domain_fid, struct fi_av_attr *attr,
+	if (av->rxd_addr_table)
+		free(av->rxd_addr_table);
+	if (av)
+		free(av);
```
free can handle null pointers.  A bigger concern is that there's a check for av == null, which implies that trying to access av->anything will lead to a crash

shefty commented on prov/rxd/src/rxd_av.c:
```diff
@@ -377,20 +378,22 @@ int rxd_av_create(struct fid_domain *domain_fid, struct fi_av_attr *attr,
 	av = calloc(1, sizeof(*av));
 	av->fi_addr_table = calloc(1, attr->count * sizeof(fi_addr_t));
```
We need to check for av != null prior to accessing it here.

aingerson commented:
@shefty Updated to include the free() fix and also to include on more patch (last one I hope...)

aingerson commented:
@shefty  Do you have any more comments or is it ok to merge?

shefty commented:
Is additional coordination needed with the fabtests patches from @a-ilango ?

aingerson commented:
I believe he is going to update his patch to exclude the revert and test_config/verbs edits but we can get his thumbs up just in case.

shefty commented on fabtests/ubertest/uber.c:
```diff
@@ -328,6 +328,55 @@ static int ft_recv_test_info(void)
+	//check needed to skip utility providers, unless requested
+	skip = (!ft_util_name(hints->fabric_attr->prov_name, &len) &&
+		strcmp(hints->fabric_attr->prov_name,
+		info->fabric_attr->prov_name)) ? 1 : 0;
```
nit: can drop ? : and use boolean value directly.

shefty commented on fabtests/ubertest/uber.c:
```diff
@@ -328,6 +328,55 @@ static int ft_recv_test_info(void)
+	if (ret)
+		return ret;
+
+	return skip || remote_skip ? 1 : 0;
```
nit: can drop ? : and return boolean value directly
also extra blank line at end of function.

shefty commented on prov/rxd/src/rxd_cq.c:
```diff
@@ -323,6 +325,15 @@ static int rxd_verify_active(struct rxd_ep *ep, fi_addr_t addr, fi_addr_t peer_a
+		dlist_remove(&pkt_entry->d_entry);
+		rxd_release_tx_pkt(ep, pkt_entry);
+		ep->peers[addr].unacked_cnt--;
+		dlist_remove(&ep->peers[addr].entry);
```
I need help here.  We're removing stuff from 3 different lists.  The packet is stored on 2 lists?  Is this correct?

aingerson replied:
First dlist_remove is an extra one. I'll take it out. We're removing the rts packet from the unacked list and then removing the peer from the rts_sent_list.

shefty commented on prov/rxd/src/rxd_ep.c:
```diff
@@ -885,16 +912,39 @@ static void rxd_peer_timeout(struct rxd_ep *rxd_ep, struct rxd_peer *peer)
+		if (pkt_entry->retry_cnt > RXD_MAX_PKT_RETRY) {
+			rxd_peer_timeout(ep,
+				&ep->peers[rxd_get_base_hdr(pkt_entry)->peer]);
+			break;
```
Why do we break here, rather than continuing to check to rest of the packets in the list?

aingerson replied:
Progress is made per peer so if the peer times out, it removes the unacked packets and completes the pending entries in error if the peer becomes unresponsive. So there's not much of a point in continuing to progress the packets.
And same for the next comment. If we have an issue sending a packet, there's not much of a point to try the rest since they will probably also result in an error or, if they do get sent, they'll be out of order at the receive side.

shefty replied:
got it -thanks

shefty commented on prov/rxd/src/rxd_ep.c:
```diff
@@ -885,16 +912,39 @@ static void rxd_peer_timeout(struct rxd_ep *rxd_ep, struct rxd_peer *peer)
+		}
+		ret = rxd_ep_retry_pkt(ep, pkt_entry);
+		if (ret)
+			break;
```
Same here

shefty commented on prov/rxd/src/rxd_rma.c:
```diff
@@ -108,8 +111,11 @@ ssize_t rxd_generic_rma(struct rxd_ep *rxd_ep, const struct iovec *iov,
+		ret = rxd_ep_send_rts(rxd_ep, rxd_addr);
+		if (ret)
+			goto out;
+	}
```
The above 7 lines of code are duplicated 4 times above.  Can this be reduced?

aingerson commented:
Updated

shefty commented:
👍
